Add Xen multiboot kernel to bootspec and package, refactor boot options, add xen bootspec support to Limine

digitalrane opened the pull request:
This PR aims to re-introduce support for booting the multiboot binary we build as part of the Xen package so that compatible bootloaders may use the Xen bootspec extension to build BIOS-bootable Xen dom0 configurations for a NixOS system. This change does not interfere in any way with the current ability to build and boot EFI-based Xen dom0 nixos systems.

It also includes changes to the Limine bootloader module to support the Xen bootspec extension, allowing booting on BIOS-based systems.

In summary, these changes:
 - Add an uncompressed multiboot binary for the Xen dom0 kernel to the `boot` output of the `xen-<version>` package
 - Extend the existing Xen bootspec extension to include EFI, Multiboot, version and parameter information
 - Refactors the options pertaining to the dom0 kernel build and configuration to not be EFI-specific.
 - Updates the existing boot builder to use the updated bootspec extension schema when fetching the EFI binary path
 - Make some small tweaks to formatting suggested by the `nixd` language server
 - Extend the Limine install script to detect and build boot entries for the Xen dom0 kernel if the bootspec extension is present for a generation

Configuration changes summary:
 - The `virtualisation.xen.boot` configuration object has been introduced
 - Several non-EFI-specific settings have been moved under `virtualisation.xen.boot` instead of `virtualisation.xen.efi`
 - The `virtualisation.xen.biosBoot` and `virtualisation.xen.efiBoot` objects have been added to the configuration, and multiboot/BIOS-specific and EFI-specific options moved under these keys
 - No Limine configuration options are modified

Bootspec extension changes summary:
 - Added `version` to the bootspec extension to make it easier to generate menus and labels in compatible bootloader modules 
 - Renamed `xen` to `efiPath` for the EFI-specific binary in the `boot` output, and updated the boot builder to use this key
 - Added `multibootPath` for the uncompressed multiboot binary within the `boot` output of the package

Limine install script changes summary:
 - Adds a helper function to generate one or more "Xen" entries for booting a Xen hypervisor, using NixOS as dom0
 - Supports both EFI (untested on hardware) and BIOS boot (tested on hardware - Coreboot+SeaBIOS)
 - The Xen helper function is called for each generation and only produces additional multiboot entires if BIOS and/or EFI booting is configured, and the relevant Hypervisor binary is present in the bootspec extension

Packaging (build support) changes for Xen summary:
 - Post-build, decompress the gzip-compressed multiboot binary as not all bootloaders (i.e Limine) support compressed multiboot kernels
 - Add the decompressed multiboot binary to the package outputs as `multiboot`

These changes do not:
- Enable GRUB to be used for booting Xen dom0 kernels. The grub support has been deprecated and we now have Limine as a maintained option for booting BIOS-based systems.
 - Interfere with existing NixOS configurations. Any changed settings in the dom0 module have been configured with `mkRenamedOptionModule` so that users are notified on the change but their existing configuration will still work with the previous EFI-specific configuration options.
 - Remove any existing functionality or configuration options. All renamed options have counterparts in the non-EFI-specific configuration schema proposed in this PR.
 - Modify any Limine module options, or interfere with existing menu generation code. All Xen code is in a separate function called when the Xen bootspec extension is present

Things I have tested:
 - End-to-end build of a BIOS-based Xen dom0 system using this module and the modified Limine install script, running on Coreboot+SeaBIOS, including general Xen functionality on the booted system. No problems were found.

**Limine Install script note!**

This one is mostly only interesting for Limine maintainers.

For some odd reason, I noticed that the first entry in the cmdline and module_string seems to be ignored when using the multiboot module. I've worked around this in the script, but also want to investigate if this is a bug in the script or the limine host binary. It seems from the logs it is being correctly passed, but when booting the first argument seems to be ignored or not received by the relevant multiboot kernel or Linux kernel (which is loaded as the second module, the Xen kernel being the first). I've worked around this by adding a -- as the first argument, as this is a no-op and everything works correctly with this in place.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

SigmaSquadron commented on nixos/modules/virtualisation/xen-dom0.nix:
```diff
@@ -51,21 +57,25 @@ let
 
     text = readFile ./xen-boot-builder.sh;
+
+    efiMountPoint = config.boot.loader.efi.efiSysMountPoint;
```
```suggestion
```
We already add this on `runtimeEnv`.

digitalrane replied:
Not sure how this crept in. Removed.

SigmaSquadron commented on nixos/modules/virtualisation/xen-dom0.nix:
```diff
@@ -51,21 +57,25 @@ let
         ]
       );
     runtimeEnv = {
+      biosBoot = cfg.biosBoot.enable;
+      efiBoot = cfg.efiBoot.enable;
```
```suggestion
```
Those variables aren't doing anything. The `systemd-boot` builder doesn't call for them.

digitalrane replied:
Yep, I elaborated on this below, but this was cruft from an initial implementation using the boot builder script before I moved to the bootspec approach once I proved that was viable. Removed.

SigmaSquadron commented on nixos/modules/virtualisation/xen-dom0.nix:
```diff
@@ -146,6 +156,61 @@ in
+        "builderVerbosity"
+      ]
+    )
+    (mkRenamedOptionModule
+      [
+        "virtualisation"
+        "xen"
+        "efi"
+        "trace"
+      ]
+      [
+        "virtualisation"
+        "xen"
+        "boot"
+        "trace"
+      ]
+    )
```
`xen.efi.trace` was never an option. `xen.trace` is. Please leave it where it is.
```suggestion
```

digitalrane replied:
Ah yep, nice catch. This was cruft from when I was first experimenting with the changes to the config structure. I've removed `efi.trace`.

SigmaSquadron commented on nixos/modules/virtualisation/xen-dom0.nix:
```diff
@@ -146,6 +156,61 @@ in
+        "efi"
+        "path"
+      ]
+      [
+        "virtualisation"
+        "xen"
+        "efiBoot"
+        "path"
+      ]
```
```suggestion
      [
        "virtualisation"
        "xen"
        "boot"
        "efi"
        "path"
      ]
```
I really like the new `boot` tree of options, but having `boot`, `efiBoot` and `biosBoot` seems weird. Could you move them to `boot.efi` and `boot.bios` respectively?

digitalrane replied:
Done, I agree that's much nicer.

SigmaSquadron commented on nixos/modules/virtualisation/xen-dom0.nix:
```diff
@@ -178,25 +243,31 @@ in
+        example = true;
+        description = "Whether to enable Xen debug tracing and logging for Domain 0.";
+      };
+      builderVerbosity = mkOption {
```
Seems like only the `systemd-boot` builder supports this? (Haven't seen anything related in the limine script) If so, please add something in the description saying that this only affects `systemd-boot` users.

digitalrane replied:
It's a little-bit indirect how this works for other bootloaders.

I just rechecked the code, and basically, here's the flow:
 - `config.trace` gets set to `true` (the default being false)
 - if `config.trace` is true `config.xenstored.log.level` gets set to `debug`
 - if `config.trace` is true, `config.boot.params` has `loglvl=all` and `guest_loglvl=all` added
 - the bootspec extension has a `params` key now which gets populated from `config.boot.params`
 - bootloaders which support the bootspec (in this case Limine) populate the multiboot parameters from the `params` key in the bootspec

So the TL;DR is that setting the `trace` flag should add the `loglvl=all` and `guest_loglvl=all` flags to the `cmdline` in the Limine boot config, as well as the original behaviour of the boot-builder script (for systemd-boot/EFI).

It also influences `xenstored` logging as well.

digitalrane replied:
`xen.trace` has been left the way it is, I was just explaining how it works currently. `xen.trace` influences the `boot.params` value currently, that's how it adds the necessary commandline flags to the Xen kernel to enable debug logging. All this change does, is make sure the bootspec has access to the params. Adding the `params` to the the bootspec is the only change this PR makes which interacts with this value.

SigmaSquadron commented on nixos/modules/virtualisation/xen-dom0.nix:
```diff
@@ -220,7 +291,15 @@ in
+    efiBoot = {
+      enable = mkOption {
+        type = bool;
+        default = true;
```
I don't know about this. How are you setting up the options for booting on BIOS systems? Do you have to explicitly disable this on BIOS systems?

digitalrane replied:
Nice catch, this was cruft from an initial implementation I'd forgotten to remove.

So, this is only really to cater to bootloaders which support both BIOS & Limine booting for Xen. Namely, Limine. I suspect if we wanted to support GRUB we'd have a similar issue. My initial implementation was going to try and handle this by extending the boot builder script, and this is left-over from this. I've pushed a commit to remove it, and the asserts which were validating this config. It's not needed.

SigmaSquadron replied:
Ideally you don't have to set up any additional settings for booting Xen, other than `virtualisation.xen.enable`. The module implicitly figures out your bootloader and hooks the Xen boot builders into the appropriate module. (Except for Lanzaboote, which as far as I can tell, checks for Xen during runtime instead of during the module system compilation.)

SigmaSquadron commented on nixos/modules/virtualisation/xen-dom0.nix:
```diff
@@ -610,10 +709,12 @@ in
-          || (config.boot ? lanzaboote) && config.boot.lanzaboote.enable;
-        message = "Xen only supports booting on systemd-boot or Lanzaboote.";
+        assertion = cfg.efiBoot.enable || cfg.biosBoot.enable;
+        message = "Either efiBoot.enable or biosBoot.enable must be enabled";
```
```suggestion
        message = "Only one of 'virtualisation.xen.boot.efi.enable' or 'virtualisation.xen.boot.bios.enable' must be enabled.";
```
Shouldn't this be "***only one of 'virtualisation.xen.boot.efi.enable' or 'virtualisation.xen.boot.bios.enable' must be enabled*"? The assertion will allow for someone enabling *both* BIOS boot and EFI boot.

digitalrane replied:
That was initially an intended configuration option yes, to cater to bootloaders capable for booting both, for users who might be migrating from bios to EFI booting or CSM to pure EFI booting. I didn't end up taking the route where this was managed by this module, and install relegated this to the bootloader module by using a pure bootspec extension approach in this PR. So this has been deleted, I've pushed a commit.

SigmaSquadron commented on pkgs/build-support/xen/default.nix:
```diff
@@ -254,8 +257,8 @@ stdenv.mkDerivation (finalAttrs: {
     cp -prvd dist/install/etc $out
+    find dist/install/boot -type f -name '*.gz' -print -exec gunzip -k '{}' ';'
     cp -prvd dist/install/boot $boot
-
```
Very nitpicky, but `runHook` calls are usually separated by newlines (and the `preInstall` call is separated by a newline above)

digitalrane replied:
TIL, thanks for the suggestion.

SigmaSquadron commented on pkgs/build-support/xen/default.nix:
```diff
@@ -254,8 +257,8 @@ stdenv.mkDerivation (finalAttrs: {
     mkdir --parents $out $out/share $boot
     cp -prvd dist/install/nix/store/*/* $out/
     cp -prvd dist/install/etc $out
+    find dist/install/boot -type f -name '*.gz' -print -exec gunzip -k '{}' ';'
```
```suggestion
    # Unzip the compressed multiboot image so Limine picks it up in the module.
    find dist/install/boot -type f -name '*.gz' -print -exec gunzip -k '{}' ';'
```
Because I *will* forget what this does.

SigmaSquadron reviewed (commented):
Seems really good already. Most of my comments are easy to resolve nitpicks.

lzcunt commented:
> For some odd reason, I noticed that the first entry in the cmdline and module_string seems to be ignored when using the multiboot module. I've worked around this in the script, but also want to investigate if this is a bug in the script or the limine host binary. It seems from the logs it is being correctly passed, but when booting the first argument seems to be ignored or not received by the relevant multiboot kernel or Linux kernel (which is loaded as the second module, the Xen kernel being the first). I've worked around this by adding a -- as the first argument, as this is a no-op and everything works correctly with this in place.

The `limine` host tool does not parse config files, so it can't be an issue there. It could be a Limine (bootloader) bug or a Xen kernel bug. Or maybe it's intended for whatever reason, I don't really know much about Xen. Will look into this soon, and will review in a few hours hopefully.

digitalrane commented:
> > For some odd reason, I noticed that the first entry in the cmdline and module_string seems to be ignored when using the multiboot module. I've worked around this in the script, but also want to investigate if this is a bug in the script or the limine host binary. It seems from the logs it is being correctly passed, but when booting the first argument seems to be ignored or not received by the relevant multiboot kernel or Linux kernel (which is loaded as the second module, the Xen kernel being the first). I've worked around this by adding a -- as the first argument, as this is a no-op and everything works correctly with this in place.
> 
> The `limine` host tool does not parse config files, so it can't be an issue there. It could be a Limine (bootloader) bug or a Xen kernel bug. Or maybe it's intended for whatever reason, I don't really know much about Xen. Will look into this soon, and will review in a few hours hopefully.

Thanks, that would be awesome. And just to clarify (I compressed some detail), this happens when the multiboot module is used, and impacts both the Linux kernel commandline and Xen kernel. In the logs however it does seem like the bootloader is parsing the file correctly and passing all the values along, they just don't seem to make it to the module with all options intact.

SigmaSquadron commented on nixos/modules/virtualisation/xen-dom0.nix:
```diff
@@ -609,12 +664,6 @@ in
         assertion = pkgs.stdenv.hostPlatform.isx86_64;
         message = "Xen is currently not supported on ${pkgs.stdenv.hostPlatform.system}.";
       }
-      {
-        assertion =
-          config.boot.loader.systemd-boot.enable
-          || (config.boot ? lanzaboote) && config.boot.lanzaboote.enable;
-        message = "Xen only supports booting on systemd-boot or Lanzaboote.";
-      }
```
Please re-add and expand this assertion to include Limine.

digitalrane replied:
I've re-added and updated the assertion

lzcunt commented on nixos/modules/system/boot/loader/limine/limine-install.py:
```diff
@@ -122,18 +131,72 @@ class BootSpec:
 def bootjson_to_bootspec(bootjson: dict) -> BootSpec:
     specialisations = bootjson['org.nixos.specialisation.v1']
     specialisations = {k: bootjson_to_bootspec(v) for k, v in specialisations.items()}
+    xen = bootjson['org.xenproject.bootspec.v1']
```
This breaks non-Xen systems because the Xen bootspec does not always exist

lzcunt replied:
Which option creates the Xen bootspec btw? I've been unable to find it and `virtualisation.xen.enable` didn't work

SigmaSquadron replied:
On Nixpkgs `master`, the Bootspec extension is defined in lines 696 to 701.
https://github.com/NixOS/nixpkgs/blob/25125c6e6b12df85d808dbfac34d6e4dd1efd589/nixos/modules/virtualisation/xen-dom0.nix#L696-L701
It looks like this on the final `boot.json`:
```json
{
  "org.nixos.bootspec.v1": {
    "options": "...",
  },
  "org.xenproject.bootspec.v1": {
    "xen": "/nix/store/i3zlvw18m2irqnpdfx70ql754k9mfddb-xen-4.20.0-boot/boot/xen-4.20.0.efi",
    "xenParams": [
      "vga=ask",
      "loglvl=all",
      "guest_loglvl=all",
      "dom0_mem=4096M",
      "dom0_max_vcpus=1"
    ]
  }
}
```

SigmaSquadron replied:
> ...and `virtualisation.xen.enable` didn't work.

Could you open an issue with what exactly didn't work? Assuming you're using `systemd-boot`, it should Just Work™

lzcunt replied:
I was testing this PR with Limine, enabling Xen didn't fix the error in the install script about the Xen bootspec not existing (which also happens with Xen off)

lzcunt replied:
```patch
diff --git a/nixos/modules/system/boot/loader/limine/limine-install.py b/nixos/modules/system/boot/loader/limine/limine-install.py
index 857d686c9a7a..d0a91335acde 100644
--- a/nixos/modules/system/boot/loader/limine/limine-install.py
+++ b/nixos/modules/system/boot/loader/limine/limine-install.py
@@ -131,14 +131,16 @@ class BootSpec:
     label: str
     toplevel: str
     specialisations: Dict[str, "BootSpec"]
-    xen: XenBootSpec
+    xen: XenBootSpec | None
     initrd: str | None = None
     initrdSecrets: str | None = None
 
 def bootjson_to_bootspec(bootjson: dict) -> BootSpec:
     specialisations = bootjson['org.nixos.specialisation.v1']
     specialisations = {k: bootjson_to_bootspec(v) for k, v in specialisations.items()}
-    xen = bootjson['org.xenproject.bootspec.v1']
+    xen = None
+    if 'org.xenproject.bootspec.v1' in bootjson:
+        xen = bootjson['org.xenproject.bootspec.v1']
     return BootSpec(
         **bootjson['org.nixos.bootspec.v1'],
         specialisations=specialisations,
```

digitalrane replied:
Thanks, I've pushed a commit to address this with the suggested solution.

lzcunt commented on nixos/modules/system/boot/loader/limine/limine-install.py:
```diff
@@ -122,18 +131,72 @@ class BootSpec:
+        # the final module is the initrd
+        entry_footer += "module_path: " + get_kernel_uri(bootspec.initrd) + "\n"
+    # load Xen dom0 as the executable, generate entries for EFI and BIOS boot
+    if config("efiSupport") and "xenEfi" in boot_spec.xen:
+        entry += entry_header
+        entry += "protocol: multiboot2\n"
+        entry += "path: " + get_copied_path_uri(bootspec.xen["xenEfi"], "xen") + "\n"
+        entry += entry_footer
```
xenEfi is supposed to be a EFI executable, eh? Why are we using multiboot2 to boot it? Maybe try `protocol: efi`? I am currently rebuilding my system to test if multiboot2 works then I'll try efi chainloading if it doesn't (which is what I expect)

digitalrane replied:
Unfortunately I don't have hardware free for this, but yes it's possible it should be the multiboot binary, or the EFI UKI directly as an EFI payload, I'm not sure what will work best. I elaborated below on the usage of multiboot1 however, as multiboot2 did not work under Limine for the Xen multiboot binary when BIOS booting.

digitalrane replied:
After some rummaging this morning I've turned up an ASRock board with a Ryzen CPU which has stock UEFI firmware. I'll test it out and update the PR based on what works. If the multiboot executable works as-is I'll be able to simplify this code significantly, if it needs the EFI executable or protocol: EFI then I'll update the script to accommodate that.

lzcunt commented on nixos/modules/system/boot/loader/limine/limine-install.py:
```diff
@@ -122,18 +131,72 @@ class BootSpec:
+        entry += "protocol: multiboot2\n"
+        entry += "path: " + get_copied_path_uri(bootspec.xen["xenEfi"], "xen") + "\n"
+        entry += entry_footer
+    elif config("biosSupport") and "xenMultiboot" in boot_spec.xen:
+        entry += entry_header
+        entry += "protocol: multiboot\n"
+        entry += (
+            "path: " + get_copied_path_uri(bootspec.xen["xenMultiboot"], "xen") + "\n"
+        )
+        entry += entry_footer
```
And why is xenMultiboot using multiboot1? Does Xen support both or does it only support multiboot1? Does xenMultiboot require a BIOS system? We can boot multiboot protocol kernels on EFI systems too.

Please not that `!efiSupport == biosSupport`, one of them has to be enabled but both of them can be enabled at the same time too - in which case this conditional has a logic error and will result in an unbootable configuration when booted with CSM on. I suggest booting either with multiboot all the time (regardless of the system being BIOS or EFI) or booting with multiboot on !biosSupport else booting with xenEfi on efiSupport (!biosSupport && !efiSupport is an unsupported unbootable configuration right now anyways).

TL;DR: we should pick (either always use or prefer) the "lowest common protocol" (which is multiboot because it can be used regardless of firmware) to not result in unbootable systems.

SigmaSquadron replied:
Xen supports multiboot2 since version 4.9. The multiboot kernel can also be booted in UEFI systems. In fact, that's the preferred boot method according to upstream.

lzcunt replied:
Then we can drop xenEfi in the Limine script.

digitalrane replied:
I mentioned I haven't specifically tested EFI booting with this config as I don't have hardware handy - so I don't have a concrete answer on the best approach for EFI. The multiboot2 support should work, and should work for EFI, but it does not work for BIOS booting and the multiboot1 module needs to be used, as the multiboot2 module complains about a missing requried section. The specific message & case I was hitting is [here](https://github.com/limine-bootloader/limine/blob/v9.x/common/protos/multiboot2.c#L216) in the Limine code. It's also possible we could just EFI boot the combined EFI image (the UKI) instead of using multiboot, but it was my understand we should be able to multiboot on EFI as well. It's also possible the multiboot1 module would work as well under EFI, looking at the code. I'd definitely appreciate some guidance on this for EFI, and if anyone has hardware, the testing would be helpful.

digitalrane replied:
Specifically, using multiboot2 under BIOS booting (the only thing I was able to test), produces this message:
`multiboot2: required EFI AMD64 entry tag is unsupported and no alternative entry was found`

lzcunt commented on nixos/modules/system/boot/loader/limine/limine-install.py:
```diff
@@ -113,6 +113,15 @@ def get_kernel_uri(kernel_path: str) -> str:
+class XenBootSpec:
+    """Represent the bootspec extension for Xen dom0 kernels"""
+
+    xenEfi: str
+    xenMultiboot: str
+    xenVersion: str
+    xenParams: List[str]
```
```suggestion
    efiPath: str
    multibootPath: str
    params: List[str]
    version: str
```
You updated the keys in the Xen bootspec but forgot to update the Limine install script, please update the rest of the script accordingly as this makes it not generate the Xen entries.

digitalrane replied:
Thanks, I did miss these when merging the two PRs back together. I've pushed a commit to make them consistent again, and also changed double quotes to single quotes in the changes, with the exception of the docstrings, or where double quotes have been used as part of log messages with the intent of the double quotes being present in the log messages themselves.

lzcunt reviewed (changes requested):
Nitpick: please try to follow the coding style used in the file - e.g. even if it's kinda mixed most of the file uses single quotes for strings.

Don't forget to fix the requested changes on the previous review too.

lzcunt commented:
The fixup commits should be squashed - `git-rebase -i HEAD~6` and then move the limine fixup commits right after the original Limine commit and mark them as fixup, do the same for the Xen fixup commit.

Not sure about Xen, but Limine commit messages should be labeled `nixos/limine:`.

UEFI should be tested. I can't do that because my configuration is incompatible with Xen (I use the old stage-1 initrd). We can likely get away with dropping the Xen EFI executable and never using it in the Limine script.

To my eyes, the multiboot2 issue in Limine looks like an unimplemented feature in Limine, I'll report that upstream. We can use multiboot1 until it's fixed.

digitalrane commented:
> The fixup commits should be squashed - `git-rebase -i HEAD~6` and then move the limine fixup commits right after the original Limine commit and mark them as fixup, do the same for the Xen fixup commit.
> 
> Not sure about Xen, but Limine commit messages should be labeled `nixos/limine:`.
> 
> UEFI should be tested. I can't do that because my configuration is incompatible with Xen (I use the old stage-1 initrd). We can likely get away with dropping the Xen EFI executable and never using it in the Limine script.
> 
> To my eyes, the multiboot2 issue in Limine looks like an unimplemented feature in Limine, I'll report that upstream. We can use multiboot1 until it's fixed.

Would the preference be to have the following?
 - The Limine install script changes in a single commit, with `nixos/limine:` as the commit prefix
 - The Xen module & build support changes in a single commit, with `nixos/xen:` as the commit prefix
 
 I originally had this as two PRs, one for the Xen + BootSpec changes and the other for the Limine changes to avoid this kind of thing, but if there's an established preference for how to handle it, I'm happy to re-order and squash the commits into whichever configuration is typical?

digitalrane commented:
TL;DR - I got Xen EFI booting, it currently requires the EFI protocol and the EFI binary for reasons I elaborate on below. Both BIOS and UEFI booting works with this PR, and I've tested it on physical hardware which supports both BIOS (via CSM) and UEFI booting, using Limine as the bootloader, and both produce a working Xen hypervisor with a NixOS dom0. I've also cleaned up the commits and split them into `nixos/xen:`, `nixos/limine:` and `xen:` commits encapsulating the relevant NixOS module changes being made, and the Xen package changes being made, into their respective commits.

In detail:

I managed to find an AMD system in my workshop with a stock UEFI implementation, so spent most of the day working out and testing how to get Xen EFI booting with Limine, and making changes to the install script to do this in the cleanest way that worked. So, I've pushed some changes to the script to be able to build per-generation directories for Xen, with the required files and configuration files for EFI-booting, and added cleanup of this directory on each run of the script as well. I've simplified the config entry code a little, and also added a helper funcition for dealing with the complexities of Xen EFI, which is only called when there is a Xen bootspec extension present for a system generation, and if `efiSupport` is enabled. I have tested this code and have been able to boot Xen with a NixOS dom0 under a standard UEFI firmware via Limine using this PR. There didn't seem to be any issues with the resulting system or Xen functionality, either.

I also found a bug related to multi-generations and generated Limine configurations for Xen. As the files were being produced in `/boot/limine/xen` for all generations, if the Xen artifacts changed without the filenames changing, the generated hashes in the Limine configuration file would match in some generations, but not in others. Typically, the newer generations would not match, based on generation build order. As part of enabling EFI booting (which needs the files and config file in the same directly) I moved to using a per-generation directory for Xen related artifacts, to prevent this collision and allow EFI booting without needing to produce a UKI for Xen.

Ideally, we would just use `multiboot2` for both EFI and BIOS booting, which would further simplify the menu generation, and remove the need for the EFI artifacts and EFI configuration files for Xen to enable EFI booting. Per Xen documentation this should work for both BIOS and UEFI. I believe this is currently possible with GRUB2, also.

On systems with both `biosSupport` and `efiSupport`, both entries will be generated. If the system supports both UEFI and BIOS via CSM or similar like my test system does, I have tested booting both the BIOS (CSM) Limine install in the MBR, and Xen via the generated BIOS entries, and also tested the UEFI Limine install, and booting the EFI Xen entries generated, and both work without any issues. Either `efiSupport` or `biosSupport` can be disabled to disable either generation of the Xen EFI and general Xen multiboot menu entries respectively. The only issue with both enabled is that Limine doesn't hide EFI entries on BIOS-booted Limine, so those entries will show up but won't boot. If they are the default (first) entry, that would mean the next entry would need to be selected. I've opened an upstream issue for this.

Relevant Xen docs:
 - https://xenbits.xen.org/docs/unstable/hypervisor-guide/x86/how-xen-boots.html
 - https://wiki.xenproject.org/wiki/Xen_EFI
 - https://xenbits.xen.org/docs/unstable/misc/efi.html

Relevant upstream issues:
 - https://github.com/limine-bootloader/limine/issues/482
 - https://github.com/limine-bootloader/limine/issues/483
 - https://github.com/limine-bootloader/limine/issues/484

lzcunt commented:
> Would the preference be to have the following?
>  - The Limine install script changes in a single commit, with `nixos/limine:` as the commit prefix
>  - The Xen module & build support changes in a single commit, with `nixos/xen:` as the commit prefix
>  
>  I originally had this as two PRs, one for the Xen + BootSpec changes and the other for the Limine changes to avoid this kind of thing, but if there's an established preference for how to handle it, I'm happy to re-order and squash the commits into whichever configuration is typical?

Number of comments aren't that important as long as it's sensical, you can create as many commits as necessary but fixup commits (e.g. `limine: Fix typo`) are not preferred. `nixos/limine:` prefix is needed yes, similarly for Xen probs. The Xen changes (at least the bootspec changes) should be applied before Limine changes.

digitalrane commented:
I've also retested this PR after making all of these changes on my original test machine (Coreboot+SeaBIOS) for pure BIOS boot  testing, and everything works correctly, I'm able to boot and use Xen on that machine using this PR.

lzcunt reviewed (approved):
Can't test Xen because of configuration incompatabilities, but it produces a sensible Limine configuration with Xen enabled and does not break non-Xen systems.

lzcunt commented:
Limine 9.2.2 is about to be released with the multiboot2 issues fixed, consider rebasing and dropping the EFI chainloading stuff from the Limine install script after 9.2.2 is in master (can't create a PR for that today or in the next few days but if r-ryantm is fast about it I can test and merge it while away from home, I have a testing setup on my phone lol). I'm fine with keeping EFI chainloading if multiboot2 still causes problems.

digitalrane commented:
I'm going to test on Limine 9.2.2 and see if we can just use multiboot2 for everything, which would be ideal. If so, I'll adapt the PR to use that approach and drop the EFI booting, and switch the BIOS boot path to use multiboot2 as well.

digitalrane commented:
I've tested against Limine 9.2.2, and whilst I am able to use multiboot2 in BIOS mode now without any issues, there seems to be some issue causing Xen's multiboot kernel to hang when booted via multiboot2 in UEFI Limine. I'm going to dig into it a little more, however I don't get serial output or any other signs that the Xen hypervisor has booted at all in EFI+Multiboot2 xen boots. The original error around not being able to find a load address however is resolved, even on EFI, so it's a step further.

digitalrane commented:
@phip1611 - thank you, that's really helpful. I did find that issue earlier when I was seeing if there were any multiboot2+uefi related issues already - and it led me to find the `CONFIG_EFI_SET_VIRTUAL_ADDRESS_MAP` Kconfig setting in Xen which I'm going to try and enable and see if that helps. Thanks for the suggestion!

CertainLach commented on nixos/modules/virtualisation/xen-dom0.nix:
```diff
@@ -692,17 +748,19 @@ in
         "org.xenproject.bootspec.v1" = {
-          xen = cfg.efi.path;
-          xenParams = cfg.bootParams;
+          efiPath = cfg.boot.efi.path;
```
This breaks lanzaboote, as it relies on the old ("xen") name.

This is an out of tree package, and I'll update it to support new name, but might be annoying...

CertainLach replied:
Also, when switching to older generation, older bootloader installation script is unable to create xen entry for the new generations, because they don't have the `xen` field.

CertainLach reviewed (approved):
Can't say anything about limine changes, but have switched my coreboot payload from edk2+systemd-boot to heads, written a custom boot generator script, and it have successfully booted build xen-qubes multiboot image

https://gist.github.com/CertainLach/d3a5e42500c58451544c363f7265bea1

Thanks!

SigmaSquadron commented:
Might be a good idea to keep the xen option pointing to efiBoot (using a recursive set in the bootspec definition) and adding a comment saying it'll be deprecated in 25.11.

digitalrane commented:
@CertainLach - thanks so much for testing that out! And yeah I agree with @SigmaSquadron's suggestion. I think there's a few options but retaining the existing `xen` option won't break anything and is pretty easy. Other options I can think of would be:

 - Updating everything that uses `xen` as a fallback if `xenEfi` doesn't exist (to address the issue mentioned about old generations), so both versions of the bootspec could be supported (very compatible, but not very clear behaviour)
 - Bumping the xen bootspec extension version and publishing two extensions with different schemas, then deprecating the old one (other than being a bit more clear, I don't think this has much benefit over just adding the `xen` field in a recursive set and setting it based on `xenEfi`) 

On the multiboot2+efi testing front, I've tried a bunch of things on this board I have with EFI, and I've actually realised for some reason on this machine with 9.2.2 I can't boot anything (even a straight Linux kernel + initramfs) when booting via EFI, so I'll mess around some more. At least, it doesn't seem Xen specific and I can stop comparing multiboot2 header definitions between Xen, Grub and Limine for now 😂

CertainLach commented:
> Updating everything that uses xen as a fallback if xenEfi doesn't exist (to address the issue mentioned about old generations), so both versions of the bootspec could be supported (very compatible, but not very clear behaviour)

Bootloader generators in older nixos generations only support "xen" attribute, so it needs to be preserved in new generations bootspecs for some time.

So this attribute has to stay until everyone moves past, and no longer needs to rollback their system.

SigmaSquadron commented:
>  • Updating everything that uses `xen` as a fallback if `xenEfi` 
> doesn't exist (to address the issue mentioned about old generations), 
> so both versions of the bootspec could be supported (very compatible, 
> but not very clear behaviour)
Nope. Too much of a hassle.

>  • Bumping the xen bootspec extension version and publishing two 
> extensions with different schemas, then deprecating the old one (other 
> than being a bit more clear, I don't think this has much benefit over 
> just adding the `xen` field in a recursive set and setting it based on 
> `xenEfi`)
Good idea! I forgot that Bootspec had a version in the extension name.
Bump the current extension to v2 and copy the old extension as v1. Leave 
a comment above the v1 copy saying it'll be deprecated in 25.11, and update
the systemd-boot builder script to use the v2 paths.

digitalrane commented:
False alarm on the system I'm testing on not being able to boot anything via EFI, I just had to tweak some kernel commandline options and my non-Xen generations boot just fine under EFI Limine 9.2.2, so there's still a Xen + multiboot2 + EFI issue to shake out, or sticking with the current EFI generation code and using efi protocol is the other option.

I'll also update this to bump the bootspec version, and make sure both bootspec versions are published by the module with a deprecation notice.

digitalrane commented:
OK, so Xen won't boot from Limine in EFI mode when using the multiboot2 protocol because Xen requires that the EFI Boot Service is still present (it hasn't been exited by the bootloader). It sets a [special tag](https://www.gnu.org/software/grub/manual/multiboot2/multiboot.html#EFI-boot-services-tag) in the multiboot2 header to request this behaviour, which GRUB has specific support for. If it sees this tag, it passes the original EFI memory map to Xen and doesn't exit the EFI Boot Service. Xen then uses the original memory map and works out a bunch of memory mapping stuff as it has more detail about how things are mapped, and then exits the EFI Boot Service itself. Limine doesn't support this behaviour, does not have support for this option tag, and exits the EFI Boot Service prior to creating a memory map to be passed to the Multiboot payload, so it won't work unless this Multiboot2 tag is supported by Limine.

I took a look at what would be needed to add support for this to Limine and it's non-trivial, especially after reading [the code in GRUB](https://git.savannah.gnu.org/cgit/grub.git/tree/grub-core/loader/multiboot_mbi2.c#n810) which handles this behaviour (looking at the `keep_bs` variable is a good way to follow it). I'll file an issue upstream with Limine, but I'd also totally understand if this is not something the Limine developers want to support. It's a fairly big workaround created in the Multiboot2 spec for enabling Xen to get access to the memory information it needs under EFI when booted via multiboot2.

For this PR, I'm going to recommend we stick with the current EFI protocol code and I'll file that issue upstream.

digitalrane commented:
Upstream issue for the EFI Boot Service tag support: https://github.com/limine-bootloader/limine/issues/490

digitalrane commented:
@lzcunt @CertainLach @SigmaSquadron - I've reworked this PR slightly to cover a few topics:
 - The new multiboot fields in the Xen bootspec extension have been moved to a v2 extension (with v1 to be deprecated in the future).
 - The EFI boot builder for systemd-boot supports both v1 and v2 bootspec extensions for  backwards & forwards compatibility with generations built with earlier versions of the module to faciliate rollback etc.
 - Given multiboot2 didn't work out for Xen, this PR is using the efi protocol for Limine, and generates the files. The Limine install script has been updated to use the v2 extension so it has the details required for multiboot and EFI booting

digitalrane commented:
Also, once https://github.com/NixOS/nixpkgs/pull/395866 is merged, I'll push the commit I've got which uses Limine 9.2.2 and switches the `multiboot` entry for BIOS booting over to using `multiboot2`, as that works on Limine 9.2.2.

programmerlexi commented:
Are you sure you spelled that last one correctly?

digitalrane commented:
I'd introduced a few bugs when adapting the systemd-boot builder for Xen to handle the v1 & v2 bootspec detection so that newer and older generations could be rebuild using  the script. I've fixed those up through a few iterations and those are all fixed now.

I've also done further testing on the current version of this PR, on real hardware, and the following works:
 - EFI via systemd-boot using the generated UKI, booted and tested on my UEFI test machine
 - EFI via Limine using the Xen EFI binary + the generated Xen efi configuration in the Limine install script, booted and tested on my UEFI test machine
 - BIOS via Limine using the Xen multiboot binary with the multiboot1 protocol (until Limine 9.2.2 is available), booted and tested on the CSM implementation on my UEFI test machine

So in summary, all the modes which this PR aims to support, are all booting successfully, and all new functionality has been moved to the `v2` bootspec extension.

Systems with `v1` bootspec extensions in the bootspec (boot.json file) for a given previous generation will still be able to use this code to move back & forth between newer and older versions of the module when using systemd-boot to load Xen.

Only system generations with a `v2` bootspec extension (and the new schema) will be able to use Limine to boot Xen via BIOS or EFI.

With this, I think all feedback & possible use cases for this PR are covered. Let me know if any questions or further feedback. I think it would make sense to wait for 9.2.2 of Limine is merged before merging this, as it'd be best to switch the `mulitboot` entry for Limine BIOS boot to `multiboot2` for future-proofing, but it's not a blocker - both will work on Limine 9.2.2 when it's merged. It would be super easy to switch to multiboot2 later as well, it'd be a single-line diff.

digitalrane commented:
@SigmaSquadron - this is ready for your re-review when you get a change. I believe all outstanding items have been addressed.

SigmaSquadron reviewed (approved):
LGTM! Just one minor nit in an option description, but that isn't blocking. If someone merges this PR before you get a chance to fix it, it's fine.

digitalrane commented on nixos/modules/virtualisation/xen-dom0.nix:
```diff
@@ -220,19 +261,33 @@ in
+          defaultText = literalExpression "\${config.virtualisation.xen.package.boot}/\${config.virtualisation.xen.package.multiboot}";
+          example = literalExpression "\${config.virtualisation.xen.package}/boot/xen-\${config.virtualisation.xen.package.version}";
+          description = ''
+            Path to the xen multiboot binary used for BIOS booting.
```
@SigmaSquadron - I've rebased and amended the commit to incorporate this change.

programmerlexi commented:
The bug of ignoring the first argument has been resolved in limine 9.2.2

digitalrane commented:
> The bug of ignoring the first argument has been resolved in limine 9.2.2 As far as i'm aware

When I tested on 9.2.2, it enabled the usage of multiboot2 successfully (which I recall being one of the expected outcomes), the first-argument issue was still present though. I suspect it will need some special handling, GRUB must be doing something different here which Xen expects which isn't explicitly documented, as I couldn't reproduce the problem with anything other than Xen in my (limited) testing. So I was going to follow up on that later and try and work out what's going on. If you've done testing though which shows some promising results, I'd be happy to re-test.

digitalrane commented:
I've rebased this PR to address the conflict with the slight refactoring which has been done in the Limine install script around specialization handling, and to pull in newer commits.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-already-reviewed/2617/2435

phaer commented:
@digitalrane Great work! Are you still interested in this, 2 months later? There's currently a merge conflict, but if you resolve that and we get a successful test run, I think it might be good to merge?

digitalrane commented:
> @digitalrane Great work! Are you still interested in this, 2 months later? There's currently a merge conflict, but if you resolve that and we get a successful test run, I think it might be good to merge?

Thank you for checking @phaer! I am still interested in getting this merged - I've just been working through some hardware issues which meant I couldn't fully test this after fixing the merge conflicts. I've got that hardware running again now so will be fixing up the merge conflicts & testing over the next few days. I'll make sure to update the PR once I've got it merging and running.

SigmaSquadron commented on nixos/modules/virtualisation/xen-dom0.nix:
```diff
@@ -692,17 +748,28 @@ in
-      # fetch the `xen.efi` path and access the `cfg.bootParams` option.
+      # fetch the dom0 kernel paths and access the `cfg.boot.params` option.
       bootspec.extensions = {
+        # Bootspec extension v1 is deprecated, and will be removed in 25.11
```
```suggestion
        # Bootspec extension v1 is deprecated, and will be removed in 26.05
```
We're in 25.11 already.

digitalrane replied:
Well, that was fast! I'm still rebasing and fixing a few rebase issues, as well as re-testing this - but I'll be sure to bump these versions.

SigmaSquadron reviewed (commented):
Looks good to me overall, but could do with some testing.

digitalrane committed:
xen: Add support for multiboot binary

digitalrane committed:
nixos/xen: Add v2 bootspec extension with multiboot support

digitalrane committed:
nixos/limine: Add support for booting Xen

JohnRTitor commented on nixos/modules/system/boot/loader/limine/limine-install.py:
Can we have a separate test for this?

For boot loaders we always want the automated tests sanity checks to pass.

digitalrane replied:
Could you please elaborate on what you are asking here @JohnRTitor? As far as I can tell there's no tests for Limine at all, and I think that would be better approached as a separate PR. This PR is just extending the existing script.

phaer replied:
> As far as I can tell there's no tests for Limine at all

There are VM tests covering limine in a bunch of scenarios: https://github.com/NixOS/nixpkgs/tree/master/nixos/tests/limine

> This PR is just extending the existing script.

Which is awesome! As you discovered below with the discrepancy of generations, there's still a risk of unintended behavior involved. Especially later on, when we update the packages or the next person extends the script. So it'd be nice to have some test coverage, even in this PR :)

There seem to be no nixos vm tests for xen yet, though? That might make it more difficult.

SigmaSquadron replied:
> There seem to be no nixos vm tests for xen yet, though?

If I may clarify: Xen is a hypervisor, and a very finicky one, at that. Any NixOS tests that boot Xen will simply cause the VM to hang because doing nested virtualisation of Xen under QEMU is extremely difficult, and likely impossible given the specific QEMU configuration NixOS tests use.

This was extensively covered in past Xen PRs, and actually implementing Xen tests would require a significant amount of work in refactoring the whole NixOS test infrastructure. I really do not think it is viable for @digitalrane to implement these tests on their own at this time. Even upstream Xen is forgoing QEMU for full hypervisor integration tests and just running it bare-metal. QEMU is only used by upstream for minor smoke tests.

JohnRTitor replied:
Okay, that's fine.

I do not use Xen, so I wouldn't probably test this PR, let me know if it works on your end, then I can check (if this broke existing normal Limine functionalities and such) and merge.

phaer replied:
Makes sense! I had hoped, somebody might have gone through the troubles of testing Xen + NixOS already, but it's true that this would be a *much* bigger task.

NixOS VM tests are, regrettably, still tightly coupled with qemu atm.  

Good to merge when limine tests pass. Running nixpgks-review atm, but I got to leave office soon and it's somewhat compute intensive as all kinds of higher-level frameworks seem to have xen support. i.e. Vagrant, multipass, docker-machine, etc :D

digitalrane replied:
Thanks @phaer for pointing me in the direction of the existing tests, that's really helpful. I'll see if there's anything useful we can do with the existing test to test the new code specifically, but yes as @SigmaSquadron has pointed out if we enable Xen we're not going to be able to test the resulting machine end-to-end. It's a large engineering task with a lot more discovery work required.

digitalrane replied:
Whilst I was developing the fix I've mentioned in a more recent comment, I looked into see if I could extend the existing Limine tests in some way which would be useful to test the parsing when Xen was enabled, and wanted to share a summary to clarify some assumptions/statements made in this thread.

- There are tests which test that Limine can be enabled in various configurations (EFI, with specializations, BIOS, etc) and that the resulting machine boots in a VM. So that's good.
- The existing tests are just validating that a machine configuration with Limine enabled in various configurations is able to boot: So in order to test this in a useful way from a Limine perspective, we would need to enable nested virtualization in the test VM in a way which satisfied Xen's hardware requirements, which as @SigmaSquadron has highlighted, is a larger engineering effort. This would enable us to test that enabling Xen and using Limine produced a bootable machine, but cannot be accomplished currently with the Limine tests alone currently.
- The current tests are testing a single generation and validating it would produce a bootable system. This would in general be useful if we could make this work for Xen, but it would not have caught the bug I found, as it was related to a single machine moving between different configurations (Xen enabled, Xen disabled). None of the current tests validate reverting configuration, so even with Xen support, bootloader testing in general could be improved to test booting, disabling or switching bootloaders, and then attempting another boot. I'm not sure at this point if the testing framework would need to be extended to cover this kind of test case, via manual manipulation of the VM state.

digitalrane commented:
OK, this is now fully rebased again, including the refactors to the Limine install script which were made recently, the move to by-name for the Xen package, and a few changes which were made to unblock Xen 4.20 - so this should merge cleanly now.

Additionally, I've tested this on my test machine and I was able to boot 4.20.1 Xen with a NixOS 25.11 dom0 successfully via Limine without any specific configuration or any issues. There's an issue with `xenstored` currently on `master` which we're looking into, but that's not related to these changes, but it did block me from testing further than booting up dom0 successfully.

digitalrane commented:
I've fixed the issue with `xenstored` in #437892 so was able to test this end-to-end fully when using the changes in that PR, and everything was working correctly in my test on my test system, using Limine as a bootloader.

digitalrane commented:
I noticed after a few more system builds that there is now a discrepancy between the generations being built with Xen and the generations being processed by Limine - so I need to look into this and most likely add an additional check to ensure that a generation being built in the Limine install script actually has Xen files and not just check for the bootspec

SigmaSquadron commented:
Opening and closing a PR will re-run the CI. If a transient failure ever happens, just close/re-open it.
